stdlib/mlang/symbolize.mc

```diff
@@ -0,0 +1,787 @@
+            let paramMap = mapFromSeq cmpString paramPairs in 
+
+            let m = mapUnion env.currentEnv.tyVarEnv paramMap in 
+            let env = updateTyVarEnv env m in 
+
+            let tyIdent = symbolizeType env def.tyIdent in
+
+            (langEnv, {ident = ident, tyIdent = tyIdent})
+        in
+        let symbSynConstructors = lam langEnv. lam synDecl. 
+            match synDecl with DeclSyn s in 
+            match mapAccumL (symbDef s.params) langEnv s.defs with (langEnv, defs) in 
+            let decl = DeclSyn {s with defs = defs} in
+            (langEnv, decl)
+        in 
+        match mapAccumL symbSynConstructors langEnv synDecls with (langEnv, synDecls) in 
+
+        -- 4. Assign names to semantic functions
+        let symbSem = lam langEnv : NameEnv. lam declSem. 
+            match declSem with DeclSem s in 
+            match setSymbol langEnv.varEnv s.ident with (varEnv, ident) in 
+
+            let langEnv = {langEnv with varEnv = varEnv} in 
+            let decl = DeclSem {s with ident = ident} in 
+        
+            (langEnv, decl)
+        in 
+        match mapAccumL symbSem langEnv semDecls with (langEnv, semDecls) in 
+
+        -- 5. Assign names to semantic bodies, params, and types
+        let symbSem2 = lam langEnv : NameEnv. lam declSem. 
+            match declSem with DeclSem s in 
+
+            let env = updateEnv env langEnv in
+
+            match symbolizeTyAnnot env s.tyAnnot with (tyVarEnv, tyAnnot) in 
+            let env = updateTyVarEnv env tyVarEnv in 
+
+            let symbArgTy = lam env : SymEnv. lam arg : {ident : Name, tyAnnot : Type}. 
+                match setSymbol env.currentEnv.varEnv arg.ident with (varEnv, ident) in 
+                let env = updateVarEnv env varEnv in 
+
+                match symbolizeTyAnnot env arg.tyAnnot with (tyVarEnv, tyAnnot) in 
+                let env = updateTyVarEnv env tyVarEnv in 
+
+                (env, {ident = ident, tyAnnot = tyAnnot})
+            in
+            let result = match optionMap (lam a. mapAccumL symbArgTy env a) s.args with Some (env, args) 
+                         then (env, Some args) else (env, None ()) in 
+            match result with (env, args) in 
+
+
+            let symbCases = lam cas : {pat : Pat, thn : Expr}. 
+                match symbolizePat env (mapEmpty cmpString) cas.pat with (thnVarEnv, pat) in
+                let varEnv = mapUnion env.currentEnv.varEnv thnVarEnv in 
+                let thn = symbolizeExpr (updateVarEnv env varEnv) cas.thn in
+                {pat = pat, thn = thn}
+            in
+            let cases = map symbCases s.cases in
+
+            let decl = DeclSem {s with cases = cases, 
+                                       tyAnnot = tyAnnot,
+                                       args = args} in 
+
+            (langEnv, decl)
+        in
+        match mapAccumL symbSem2 langEnv semDecls with (langEnv, semDecls) in 
```

Comment:
No need for a `mapAccumL` here, using a `map` makes it clearer that no names escape from the sem bodies